src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+                        }
+                        catch (RequestFailedException ex) when (ex.ErrorCode == BlobErrorCode.BlobNotFound)
+                        {
+                            _logger.LogError(new UnclaimedPartitionException($"Partition {partition} has never been claimed"));
+                            _logger.LogMetric(EventMetrics.PartitionLastClaimedDelay(partition), DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
+                        }
+                    }
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private Task RenewOwnership(CancellationToken ct)
+        {
+            Task.Run(async () =>
+            {
+                while (!ct.IsCancellationRequested)
+                {
+                    foreach (var partition in _ownedPartitions)
+                    {
+                        _logger.LogTrace($"Renewing ownership for partition {partition}");
+                        BlobClient ownershipClient = _containerClient.GetBlobClient($"ownedpartitions/{partition}");
+                        var leaseClient = ownershipClient.GetBlobLeaseClient(_processorId);
+                        await leaseClient.RenewAsync(default, ct);
+
+                        // upload to update last modified timestamp
+                        BlobUploadOptions blobUploadOptions = new BlobUploadOptions()
+                        {
+                            Conditions = new BlobRequestConditions()
+                            {
+                                LeaseId = _processorId,
+                            },
+                        };
+                        await ownershipClient.UploadAsync(BinaryData.FromString(string.Empty), blobUploadOptions);
+                    }
+
+                    await Task.Delay(20000, ct);
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private async Task StartAssignedPartitionProcessor(CancellationToken ct, List<int> ownedPartitions)
+        {
+            var maximumBatchSize = 100;
+
+            string[] partitions = new string[ownedPartitions.Count];
+            for (var i = 0; i < ownedPartitions.Count; i++)
+            {
+                partitions[i] = ownedPartitions.ElementAt(i).ToString();
+            }
+
+            var eventHubName = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubName, nameof(_eventHubClientOptions.EventHubName));
+            var eventHubNamespaceFQDN = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubNamespaceFQDN, nameof(_eventHubClientOptions.EventHubNamespaceFQDN));
+            var eventHubConsumerGroup = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubConsumerGroup, nameof(_eventHubClientOptions.EventHubConsumerGroup));
+            var tokenCrendential = EnsureArg.IsNotNull(_eventHubClientOptions.EventHubTokenCredential, nameof(_eventHubClientOptions.EventHubTokenCredential));
+
+            var processor = new AssignedPartitionProcessor(
```

Comment:
I would recommend creating an interface for this class and injecting it in the constructor rather than instantiating it here. It's best to let the IoC Container handle resolving and creating classes, rather than creating them in the class.

Reply:
I think the main limiting factor here is the processor needs the number of partitions it is working on and we don't know them at the time of DI.

Reply:
I think a factory is the better pattern here if we want to get test coverage for this code.

Reply:
added a factory